lib/services/database/challenge_repository_service.dart

```diff
@@ -0,0 +1,175 @@
+    final pastChallengesCollectionRef = parentRef
+        .collection(ChallengeFirestore.pastChallengesSubCollectionName);
+
+    final challengesSnap = await challengesCollectionRef.get();
+    final pastChallengesSnap = await pastChallengesCollectionRef.get();
+
+    final pastPostIds = pastChallengesSnap.docs
+        .map((doc) => PostIdFirestore(value: doc.id))
+        .toSet();
+
+    final List<ChallengeFirestore> activeChallenges =
+        List.empty(growable: true);
+
+    for (final challengeSnap in challengesSnap.docs) {
+      final challenge = ChallengeFirestore.fromDb(challengeSnap);
+      if (challenge.data.isExpired ||
+          !await _postRepositoryService.postExists(challenge.postId)) {
+        _deleteChallenge(challengeSnap, parentRef);
+        pastPostIds.add(challenge.postId);
+      } else {
+        activeChallenges.add(challenge);
+      }
+    }
+
+    final now = DateTime.now();
+    final sum = now.add(maxChallengeDuration);
+    final expiresOn = DateTime(
+      sum.year,
+      sum.month,
+      sum.day,
+    ); // truncates to the day
+
+    final possiblePosts = await inRangeUnsortedPosts(pos);
+    final postIt = possiblePosts.iterator;
+    final activePostIds = activeChallenges.map((e) => e.postId).toSet();
+    while (activeChallenges.length < maxActiveChallenges && postIt.moveNext()) {
+      final post = postIt.current;
+
+      if (pastPostIds.contains(post) || activePostIds.contains(post)) {
+        continue;
+      }
+
+      final newChallenge = ChallengeFirestore(
+        postId: post,
+        data: ChallengeData(
+          isCompleted: false,
+          expiresOn: Timestamp.fromDate(expiresOn),
+        ),
+      );
+
+      await challengesCollectionRef
+          .doc(post.value)
+          .set(newChallenge.data.toDbData());
+      activeChallenges.add(newChallenge);
+      activePostIds.add(post);
+    }
+
+    return activeChallenges;
+  }
+
+  /// moves the challenge with id [pid] from the active challenges to the past challenges
+  Future<void> _deleteChallenge(
+    DocumentSnapshot<Map<String, dynamic>> challengeSnap,
+    DocumentReference parentRef,
+  ) async {
+    final batch = _firestore.batch();
+    batch.set(
+      parentRef
+          .collection(ChallengeFirestore.pastChallengesSubCollectionName)
+          .doc(challengeSnap.id),
+      challengeSnap.data()!,
+    );
+    batch.delete(challengeSnap.reference);
+    await batch.commit();
+  }
+
+  Future<Iterable<PostIdFirestore>> inRangeUnsortedPosts(GeoPoint pos) async {
+    final geoFirePosition = GeoFirePoint(pos);
+    Iterable<PostFirestore> possiblePosts =
+        await _postRepositoryService.getNearPosts(
+      pos,
+      maxChallengeRadius,
+    );
+
+    possiblePosts = possiblePosts.where(
+      (post) =>
+          geoFirePosition.distanceBetweenInKm(
+            geopoint: post.location.geoPoint,
+          ) >
+          minChallengeRadius,
+    );
```

Comment:
I personally think that it would be better to add a `minRadius` optional parameter (and its documentation) to the `_postRepositoryService.getNearPosts`. I'm feeling it would be more concise, more readable, and a better separation of concerns.

Reply:
Yeah that's a good idea, I'll keep this in this pr since its a small change, and update the pr description.
I did that in fd198fc4340bbb7db339f62b8d1a5307963d0e1b and 4f7cd91af353f1fa20a73854ab213e168a134f39